core/api/service/api_service.cpp

```diff
@@ -60,91 +60,110 @@ namespace kagome::api {
-                  result["changes"] = std::move(out_data);
-                  result["block"] = api::makeValue(block);
-
-                  jsonrpc::Value::Struct params;
-                  params["result"] = std::move(result);
-                  params["subscription"] = api::makeValue(set_id);
-
-                  self->server_->processJsonData(
-                      params, [&](const std::string &response) {
-                        session->respond(response);
-                      });
-                }
-              });
-
-          session_context.events_subscription->setCallback(
-              [wp](uint32_t set_id, SessionPtr &session, const auto &key, const auto &header) {
-                if (auto self = wp.lock()) {
-                  jsonrpc::Value::Struct params;
-                  params["result"] = api::makeValue(header.get());
-                  params["subscription"] = api::makeValue(set_id);
-
-                  self->server_->processJsonData(
-                      params, [&](const std::string &response) {
+              session_context.events_subscription->setCallback(
+                  [wp](uint32_t set_id,
+                       SessionPtr &session,
+                       const auto &key,
+                       const auto &header) {
+                    if (auto self = wp.lock()) {
+                      jsonrpc::Value::Struct p;
+                      p["result"] = api::makeValue(header.get());
+                      p["subscription"] = api::makeValue(set_id);
+
+                      jsonrpc::Request::Parameters params;
+                      params.push_back(std::move(p));
+                      self->server_->processJsonData(
+                          "chain_newHead", params, [&](const auto &response) {
+                            if (response.has_value())
+                              session->respond(response.value());
+                            else
+                              self->logger_->error("ERROR => {}",
+                                                   response.error().message());
+                          });
+                    }
+                  });
+            }
+
+            session->connectOnRequest(
+                [wp](std::string_view request,
+                     std::shared_ptr<Session> session) mutable {
+                  auto self = wp.lock();
+                  if (not self) return;
+
+                  auto thread_session_auto_release = [](void *) {
+                    threaded_info.releaseSessionId();
+                  };
+
+                  threaded_info.storeSessionId(session->id());
+
+                  /**
+                   * Unique ptr object to autorelease sessions.
+                   */
+                  std::unique_ptr<void, decltype(thread_session_auto_release)>
+                  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+                  thread_session_keeper(reinterpret_cast<void *>(0xff),
```

Comment:
It looks like durty hack. I'm suggest to make class-helper is using RAII for doing some actions at quit scope.

Reply:
PR was not requested for review yet.

Reply:
I add an explanation in comment why it looks so.